797 ICD tests in the backend unit tests

markccchiang opened the pull request:
This PR is the trial for #797 to move some ICD tests back to the backend unit tests. I will add more testing cases according to the [document](https://docs.google.com/spreadsheets/d/1qwdTTKdNA1oNnNAUNgyPZH5HOVyt2Uw0vPT2_UsEaXk/edit#gid=0)).

markccchiang committed:
add a Session tester

markccchiang committed:
solve merge conflicts

markccchiang committed:
change address and icd_version strings for the TestSession

markccchiang committed:
solve a merge conflict

markccchiang committed:
modify the TestOnRegisterViewer

markccchiang committed:
add ICD test cases

markccchiang committed:
change the testing file name

markccchiang committed:
remove unused codes

markccchiang committed:
create a DummyBackend class

markccchiang committed:
add an opening file test

markccchiang committed:
fix the code style

markccchiang committed:
move the callback check functions

markccchiang committed:
modify the ANIMATOR_NAVIGATION test

markccchiang committed:
modify the ANIMATOR_NAVIGATION test

markccchiang committed:
add a function ClearMessagesQueue in the Session

markccchiang committed:
Merge branch 'dev' of https://github.com/CARTAvis/carta-backend into mark/797_ICD_tests

markccchiang committed:
fix the code style

markccchiang committed:
replace the function CheckMessagesQueue with the TryPopMessagesQueue

markccchiang committed:
move ICD testing files to the folder icd_test

markccchiang committed:
solve merge conflicts

markccchiang committed:
refactor the testing codes

markccchiang committed:
add the BackendTester class

markccchiang committed:
add the class ElapsedTimer

markccchiang committed:
add more tests to the TestAnimatorNavigation

markccchiang committed:
add a tester TestAnimatorDataStream

markccchiang committed:
add a function template

veggiesaurus commented on test/icd_test/BackendTester.h:
```diff
@@ -0,0 +1,37 @@
+
+bool FileExists(string filename);
+
+class ElapsedTimer {
```
We already have a timer class, with it's own set of unit tests.

veggiesaurus commented:
> Since the [original ICD tests](https://github.com/CARTAvis/carta-backend-ICD-test) use some image samples with large sizes, I do not include these images in the backend repo. Users need to copy these images manually to run the test

We can automatically generate these images. Please follow the implementation @confluence used.

I also think we don't have to strictly keep the same test approach, that's not the point of this migration. If something can be tested simply as a unit test (e.g. Does Carta generate the correct file information?) then there's really no need to wrap it in then entire ICD. Just test the functionality.

Some examples that would make trivial unit tests and not require any specific ICD stuff (although that will be useful for certain tests)

- histograms
- file lists
- file info
- contours
- region stats
- moment maps
- saved images

confluence commented:
Agreed -- tests which confirm that each message has an expected and well-formed response should remain as ICD tests. Tests which check the _correctness_ of the data should be removed from the ICD tests and rewritten as ordinary unit tests which bypass the messaging interface and call functions in the backend objects directly. Wherever possible we should use generated images.

markccchiang committed:
Merge branch 'dev' of https://github.com/CARTAvis/carta-backend into mark/797_ICD_tests

markccchiang committed:
fix the protobuf problems

markccchiang committed:
Merge branch 'dev' of https://github.com/CARTAvis/carta-backend into mark/797_ICD_tests

markccchiang committed:
Merge branch 'dev' of https://github.com/CARTAvis/carta-backend into mark/797_ICD_tests

markccchiang committed:
refactor the code

markccchiang committed:
add more message handler functions to the DummyBackend

markccchiang committed:
add a message counter and log message event type

markccchiang committed:
add a bool _use_tbb_task to the Session

markccchiang committed:
add an ICD tester TestRegionRegister

markccchiang committed:
add an ICD tester TestRegionRegister

markccchiang committed:
use generated images to do ICD tests

markccchiang committed:
add a Session setting parameter: use_tbb_task

markccchiang committed:
add an ICD tester TestRegionStatisticsRectangle

markccchiang committed:
remove unnecessary checks from TestRegionRegister

markccchiang committed:
add ICD testers TestRegionStatisticsEllipse and TestRegionStatisticsPolygon

markccchiang committed:
add a large image folder for more test cases

markccchiang committed:
add a tester TestRegionSpectralProfileRectangle

markccchiang committed:
Merge branch 'dev' of https://github.com/CARTAvis/carta-backend into mark/797_ICD_tests

markccchiang committed:
modify and add functions to the ProtobufInterface

markccchiang committed:
fix the bug of sending duplicate protobuf messages

markccchiang committed:
remove the bool _use_tbb_task from the Session

markccchiang committed:
apply tbb task to the DummyBackend

markccchiang committed:
fix the code style

markccchiang committed:
set the DummyBackend function WaitJobFinished() as a public function

markccchiang committed:
apply tbb task to the DummyBackend

markccchiang committed:
try to solve building failure for the Ununtu

markccchiang committed:
rename the class DummyBackend as BackendModel

markccchiang committed:
refactor the code and add a TestAnimatorPlayback

markccchiang committed:
add an animation play backward test

markccchiang committed:
minor code changes

markccchiang committed:
rename the ICD testing codes directory

markccchiang committed:
rename the ICD testing codes directory

markccchiang committed:
refactor the code and add a benchmark test

markccchiang committed:
add a benchmark test

markccchiang committed:
add a copyright header

markccchiang committed:
modify the performance tester TestOpenFiles

markccchiang committed:
modify ICD testing codes and use generated images for testing

markccchiang committed:
update icd-test and performance-test folders

markccchiang committed:
fix the code style and some minor changes

markccchiang committed:
minor changes in cmake files

markccchiang committed:
Merge branch 'dev' of https://github.com/CARTAvis/carta-backend into mark/797_ICD_tests

markccchiang committed:
remove google benchmark testers and will apply them in an another PR

markccchiang committed:
rename files and clean up codes

markccchiang committed:
remove repeated functions from the ProtobufUtilities

markccchiang committed:
remove ElapsedTimer and use carta::Timer instead

markccchiang committed:
refactor OnMessageTask and related codes in the Main.cc

markccchiang committed:
move ProtobufUtilities.h/cc and Util.h/cc to the new folder Util

markccchiang committed:
fix the code style

markccchiang committed:
fix the code style

markccchiang committed:
fix the build error for Ubuntu

markccchiang commented:
Now I tried to integrate ICD testing codes with current backend codes. The `Session` class is modified so it can remove the correlation with uWebSockets. Refactored protobuf messages handlers on `OnMessageTask` and `Main.` Decoding incoming messages are only done in the `Main.` Added a new folder `Util` and moved `Util.h/.cc` to this folder. Also, I created new `ProtobufUtilities.h/.cc` for creating or filling protobuf messages helpers. As for the rest of the ICD tests, I will continue them in another PR.

ajm-ska commented:
I get a segmentation fault when I try to run it on MacOS:

```
./carta_backend_icd_tests       
[==========] Running 14 tests from 11 test cases.
[----------] Global test environment set-up.
[----------] 1 test from TestAnimatorPlayback
[ RUN      ] TestAnimatorPlayback.ANIMATOR_PLAYBACK
2021-09-10 06:44:41	INFO	FITSCoordinateUtil::fromFITSHeader	passing empty or nonexistant spectral Coordinate axis
zsh: segmentation fault  ./carta_backend_icd_tests
```

Maybe the images are missing? 
> Users need to copy these images manually to run the test.

Is there a list of images that are required?

ajm-ska commented:
I forgot that the ordinary unit tests have a problem on MacOS. When I run `./carta_backend_tests`, sometimes they pass, sometimes I get `zsh: segmentation fault  ./carta_backend_tests`, so it is not an exclusive issue with `./carta_backend_icd_tests`. I will just test using Linux instead.

ajm-ska commented:
I have run it several times in an Ubuntu 20.04 Docker container. It is a good start, but unfortunately, just like the old ICD tests, they seem to be unable consistently pass every time. Sometimes all 14 tests pass, other times some fail. For example, I have got:

```
[ RUN      ] TestAnimatorPlayback.ANIMATOR_PLAYBACK
2021-09-10 07:44:33	INFO	FITSCoordinateUtil::fromFITSHeader	passing empty or nonexistant spectral Coordinate axis
/root/carta-backend/test/icd-test/TestAnimatorPlayback.cc:189: Failure
Expected equality of these values:
  sync_channel
    Which is: 10
  expected_channel
    Which is: 18
/root/carta-backend/test/icd-test/TestAnimatorPlayback.cc:189: Failure
Expected equality of these values:
  sync_channel
    Which is: 10
  expected_channel
    Which is: 17
/root/carta-backend/test/icd-test/TestAnimatorPlayback.cc:189: Failure
Expected equality of these values:
  sync_channel
    Which is: 9
  expected_channel
    Which is: 16
/root/carta-backend/test/icd-test/TestAnimatorPlayback.cc:189: Failure
Expected equality of these values:
  sync_channel
    Which is: 18
  expected_channel
    Which is: 15
/root/carta-backend/test/icd-test/TestAnimatorPlayback.cc:204: Failure
Expected equality of these values:
  message_count
    Which is: 76
  expected_response_messages
    Which is: 33
[  FAILED  ] TestAnimatorPlayback.ANIMATOR_PLAYBACK (680 ms)
```

```
[ RUN      ] TestRegionSpectralProfileRectangle.REGION_SPECTRAL_PROFILE_RECTANGLE
2021-09-10 07:44:43	INFO	FITSCoordinateUtil::fromFITSHeader	passing empty or nonexistant spectral Coordinate axis
/root/carta-backend/test/icd-test/TestRegionSpectralProfileRectangle.cc:135: Failure
Expected equality of these values:
  message_count
    Which is: 2
  1
[  FAILED  ] TestRegionSpectralProfileRectangle.REGION_SPECTRAL_PROFILE_RECTANGLE (15 ms)
```

```
[ RUN      ] TestAnimatorPlayback.ANIMATOR_PLAYBACK
2021-09-10 07:47:06	INFO	FITSCoordinateUtil::fromFITSHeader	passing empty or nonexistant spectral Coordinate axis
/root/carta-backend/test/icd-test/TestAnimatorPlayback.cc:189: Failure
Expected equality of these values:
  sync_channel
    Which is: 18
  expected_channel
    Which is: 19
/root/carta-backend/test/icd-test/TestAnimatorPlayback.cc:204: Failure
Expected equality of these values:
  message_count
    Which is: 17
  expected_response_messages
    Which is: 33
[  FAILED  ] TestAnimatorPlayback.ANIMATOR_PLAYBACK (663 ms)
```

I also happened to get `Segmentation fault (core dumped)` and a time when it seemingly got stuck. I guess this will be hard to debug.

Are you planning to add the --gtest command line flags like the ones we see when we do `./carta_backend_tests --help` so that we can focus on certain tests?

kswang1029 commented:
> I have run it several times in an Ubuntu 20.04 Docker container. It is a good start, but unfortunately, just like the old ICD tests, they seem to be unable consistently pass every time. Sometimes all 14 tests pass, other times some fail. For example, I have got:
> 
> ```
> [ RUN      ] TestAnimatorPlayback.ANIMATOR_PLAYBACK
> 2021-09-10 07:44:33	INFO	FITSCoordinateUtil::fromFITSHeader	passing empty or nonexistant spectral Coordinate axis
> /root/carta-backend/test/icd-test/TestAnimatorPlayback.cc:189: Failure
> Expected equality of these values:
>   sync_channel
>     Which is: 10
>   expected_channel
>     Which is: 18
> /root/carta-backend/test/icd-test/TestAnimatorPlayback.cc:189: Failure
> Expected equality of these values:
>   sync_channel
>     Which is: 10
>   expected_channel
>     Which is: 17
> /root/carta-backend/test/icd-test/TestAnimatorPlayback.cc:189: Failure
> Expected equality of these values:
>   sync_channel
>     Which is: 9
>   expected_channel
>     Which is: 16
> /root/carta-backend/test/icd-test/TestAnimatorPlayback.cc:189: Failure
> Expected equality of these values:
>   sync_channel
>     Which is: 18
>   expected_channel
>     Which is: 15
> /root/carta-backend/test/icd-test/TestAnimatorPlayback.cc:204: Failure
> Expected equality of these values:
>   message_count
>     Which is: 76
>   expected_response_messages
>     Which is: 33
> [  FAILED  ] TestAnimatorPlayback.ANIMATOR_PLAYBACK (680 ms)
> ```
> 
>  ```
> [ RUN      ] TestRegionSpectralProfileRectangle.REGION_SPECTRAL_PROFILE_RECTANGLE
> 2021-09-10 07:44:43	INFO	FITSCoordinateUtil::fromFITSHeader	passing empty or nonexistant spectral Coordinate axis
> /root/carta-backend/test/icd-test/TestRegionSpectralProfileRectangle.cc:135: Failure
> Expected equality of these values:
>   message_count
>     Which is: 2
>   1
> [  FAILED  ] TestRegionSpectralProfileRectangle.REGION_SPECTRAL_PROFILE_RECTANGLE (15 ms)
> ```
> 
>  ```
> [ RUN      ] TestAnimatorPlayback.ANIMATOR_PLAYBACK
> 2021-09-10 07:47:06	INFO	FITSCoordinateUtil::fromFITSHeader	passing empty or nonexistant spectral Coordinate axis
> /root/carta-backend/test/icd-test/TestAnimatorPlayback.cc:189: Failure
> Expected equality of these values:
>   sync_channel
>     Which is: 18
>   expected_channel
>     Which is: 19
> /root/carta-backend/test/icd-test/TestAnimatorPlayback.cc:204: Failure
> Expected equality of these values:
>   message_count
>     Which is: 17
>   expected_response_messages
>     Which is: 33
> [  FAILED  ] TestAnimatorPlayback.ANIMATOR_PLAYBACK (663 ms)
> ```
> 
> I also happened to get `Segmentation fault (core dumped)` and a time when it seemingly got stuck. I guess this will be hard to debug.
> 
> Are you planning to add the --gtest command line flags like the ones we see when we do `./carta_backend_tests --help` so that we can focus on certain tests?

@markccchiang do you use FPS=5 for the animator test? If so could you try FPS=2 instead so that minimal cpu resource is required. I see a similar problem with robot test too.

veggiesaurus reviewed (changes requested):
@markccchiang I thought we agreed that you would integrate things into the existing unit tests for this PR? And just include a single additional test in this PR that makes use of the backend model stuff?

Many of the tests here are still simply ICD tests being run on the backend instead of via the frontend. We want to migrate many of these tests to real unit tests, that just tests the functionality. For example, region stats should be just testing the actual region stats part of the code, not the full ICD.

Are there still image files required, or are you only using generated files?

markccchiang commented:
> @markccchiang I thought we agreed that you would integrate things into the existing unit tests for this PR? And just include a single additional test in this PR that makes use of the backend model stuff?
> 
> Many of the tests here are still simply ICD tests being run on the backend instead of via the frontend. We want to migrate many of these tests to real unit tests, that just tests the functionality. For example, region stats should be just testing the actual region stats part of the code, not the full ICD.
> 
> Are there still image files required, or are you only using generated files?

Sorry for that! I think I missed your point in the discussion last time. Will change the code and use a single additional test that makes use of the backend model. Now I only use generated files in ICD tests.

markccchiang commented:
@ajm-asiaa Sometimes I also get segmentation fault when running `./carta_backend_tests` or `./carta_backend_icd_tests`. The problem of `ANIMATOR_PLAYBACK` and `REGION_SPECTRAL_PROFILE_RECTANGLE` also happens in my test. I guess there is something wrong when parallelizing these computations. But in the unit test definition, as @veggiesaurus mentioned that we only test the functionality. There is no problem from this point of view. It is a good idea that adds command line flags to only run some specific tests. I will try it.

markccchiang committed:
add a frame rate setting in the animation request message

markccchiang committed:
initialize the tbb threads count in the testing environment

markccchiang committed:
fix the problem of parsing gtest arguments

markccchiang committed:
integrate ICD tests in unit tests with a single additional file

ajm-ska commented:
@markccchiang Nice work. It runs fairly well after testing on Ubuntu 20.04. Sometimes I still get the segmentation fault, even when only running the new tests e.g `./carta_backend_tests --gtest_filter=TestIcd*  --gtest_repeat=50 --gtest_shuffle`, but I know it is not related to these new tests in particular so probably out of the scope of this pull request.

I think you noticed the issue on Jenkins with MacOS last night:

```
> OMP: Error #131: Thread identifier invalid.
> OMP: Error #131: Thread identifier invalid.
> OMP: Error #131: Thread identifier invalid.
> OMP: Error #131: Thread identifier invalid.
```

I looked it up and luckily found the solution [here](https://github.com/microsoft/LightGBM/issues/4229). It is a known issue and is apparently due to a bug in Homebrew's current libomp 12.0.1 on Big Sur (MacOS 11) only. Their suggested solution is to downgrade to the older libomp version 11 for now. I have done so, and now the tests pass on MacOS CI again.

ajm-ska commented:
@markccchiang  I have one suggestion. I'm not sure what everyone else thinks? @veggiesaurus? I know the ICD test names were previously in capital letters for some reason, but would it be worth renaming them to capital and lower case (plus without underscores) so that they would be consistent with the rest of the Unit Test test names?
```
HistogramTest.
  TestHistogramBehaviour
  TestHistogramConstructor
  TestHistogramAdd
  TestSingleThreading
  TestMultithreading
TestIcd.
  ACCESS_CARTA_DEFAULT
  ACCESS_CARTA_KNOWN_DEFAULT
  ACCESS_CARTA_NO_CLIENT_FEATURE
  ACCESS_CARTA_SAME_ID_TWICE
  ANIMATOR_DATA_STREAM
  ANIMATOR_NAVIGATION
  ANIMATOR_PLAYBACK
  REGION_REGISTER
MomentTest.
  CheckConsistency
  CheckConsistencyForBeamConvolutions
```

markccchiang committed:
change the name style of ICD test to upper camel case

confluence reviewed (changes requested):
This PR now contains too many unrelated and significant changes. I think that we should split out the message wrapper class into its own PR, and review and merge that separately.

We can review that change in more detail in that PR, but in brief I would suggest that you rename the file of wrappers from `ProtobufUtilities` to `Message`, make all the wrapper functions static functions on a `Message` class, and rename them to have the same name as the message: e.g. `Message::RegisterViewer` should return a `CARTA::RegisterViewer`.

It's fine to put it in a `Util` subdirectory for now; this is consistent with the approach I'm currently taking to refactor `Util` into separate files. But please don't move `Util.h` and `Util.cc` into the subdirectory; it's an unnecessary change, and these files are being eliminated anyway.

In that PR we should also update the rest of the code to use these wrappers consistently to construct messages. This should also include replacing the three `Fill...` functions with functions which return messages.

markccchiang committed:
move Util.h/cc from the Util folder

markccchiang committed:
rename the ProtobufUtilities as Message and create a Message class

markccchiang committed:
add a Message function to fill the CARTA::Histogram message

markccchiang committed:
add a function to create the CARTA::SpectralProfileData message

markccchiang committed:
add a function to fill the CARTA::StatisticsValue message

markccchiang commented:
Convert this PR to draft. I will reopen it after the code change about the message wrapper is merged into dev.

confluence commented:
For clarity, I was also requesting that you move `Util/Message.*` and the associated changes into a separate PR.

markccchiang committed:
solve merge conflicts

markccchiang committed:
revert the defined name of Util.h

markccchiang committed:
Merge branch 'dev' of https://github.com/CARTAvis/carta-backend into mark/797_ICD_tests

markccchiang committed:
revert unrelated changes

markccchiang committed:
Merge branch 'dev' of https://github.com/CARTAvis/carta-backend into mark/797_ICD_tests

markccchiang committed:
fix building error for the carta_backend_tests

markccchiang committed:
fix the code style

markccchiang committed:
add an ICD test case: ResumeImages

markccchiang committed:
refactor the TestIcd

confluence commented on src/Session.cc:
```diff
@@ -2098,3 +2100,11 @@ void Session::CloseCachedImage(const std::string& directory, const std::string&
+
+void Session::ClearMessagesQueue() {
+    _out_msgs.clear();
+}
```
Rather than adding these functions to the main Session class, please make `_out_msgs` protected, and create a subclass of Session in the test code which has functionality that is only used from tests. This is more consistent with our current approach -- see for example `TestFrame` and `TestSimpleFrontendServer`.

markccchiang committed:
Merge branch 'dev' of https://github.com/CARTAvis/carta-backend into mark/797_ICD_tests

markccchiang committed:
replace the TBB with carta::ThreadManager in ICD testers

markccchiang committed:
create a TestSession which inherits the Session and access _out_msgs through it

confluence reviewed (approved):
I think that this is all fine now. I'm not 100% sure if we should be making *all* private members of `Session` protected, but I don't have a specific objection.

pford commented:
> I also think we don't have to strictly keep the same test approach, that's not the point of this migration. If something can be tested simply as a unit test (e.g. Does Carta generate the correct file information?) then there's really no need to wrap it in then entire ICD. Just test the functionality.
> 
> Some examples that would make trivial unit tests and not require any specific ICD stuff (although that will be useful for certain tests)
> 
>     * histograms
>     * file lists
>     * file info
>     * contours
>     * region stats
>     * moment maps
>     * saved images

I am confused by this PR and the issue comments.  This test framework is very high-level at the message received/response level, rather than as close as possible to the code which does the work.  I thought the goal was not to duplicate the ICD tests?

confluence commented:
My understanding is that we decided that it was a good idea to have one specific unit test which tests the Session's message handling (which is what this PR currently contains) and to add most of the converted tests (the original purpose of this PR) in stages, in multiple PRs, rather than adding them in one large PR. But perhaps one group of tests should be selected to be added to this PR, so that we can further discuss the approach here.

markccchiang commented:
> I am confused by this PR and the issue comments. This test framework is very high-level at the message received/response level, rather than as close as possible to the code which does the work. I thought the goal was not to duplicate the ICD tests?

I agree that some tests can just be trivial unit tests. We don't need to duplicate the original ICD tests. We can simplify and migrate some ICD tests to the backend so that we can inspect in detail why some of the original ICD tests failed, like animation and moments generation. The main purpose of high-level tests is to check if the message received/response fits our expectations (in case if we want to do this kind of test for the new feature without the frontend). I think it can also be used to test the performance in the future. I will check if some of the test cases can be removed. Or just leave it as an example for the developer who wants to add more tests (@acdo2002 seems to have the interest to do it).

markccchiang committed:
Merge branch 'dev' of https://github.com/CARTAvis/carta-backend into mark/797_ICD_tests

markccchiang committed:
remove IcdTest.ResumeImages and decrease the frame rate for animation tests

acdo2002 commented:
I have test the ICD unit tests on the Ubuntu 20.04.
(1) After the tests, the carta_backend_test stuck there, for example:
```
[==========] 7 tests from 1 test suite ran. (5523 ms total)
[  PASSED  ] 7 tests.


```
(2) Some problem with AnimatorNavigation:

> mingyi@mingyi-ASUS-ExpertCenter-D900TA-M900TA:~/carta-backend-unit/build/test$ ./carta_backend_tests --gtest_filter=IcdTest.*
> Note: Google Test filter = IcdTest.*
> [==========] Running 8 tests from 1 test suite.
> [----------] Global test environment set-up.
> [----------] 8 tests from IcdTest
> [ RUN      ] IcdTest.AccessCartaDefault
> [       OK ] IcdTest.AccessCartaDefault (1 ms)
> [ RUN      ] IcdTest.AccessCartaKnownDefault
> [       OK ] IcdTest.AccessCartaKnownDefault (0 ms)
> [ RUN      ] IcdTest.AccessCartaNoClientFeature
> [       OK ] IcdTest.AccessCartaNoClientFeature (0 ms)
> [ RUN      ] IcdTest.AccessCartaSameIdTwice
> [       OK ] IcdTest.AccessCartaSameIdTwice (0 ms)
> [ RUN      ] IcdTest.AnimatorDataStream
> 2021-11-15 07:23:31	INFO	FITSCoordinateUtil::fromFITSHeader	passing empty or nonexistant spectral Coordinate axis
> [       OK ] IcdTest.AnimatorDataStream (456 ms)
> [ RUN      ] IcdTest.AnimatorNavigation
> sh: 1: fits2idia: not found
> sh: 1: fits2idia: not found
> /home/mingyi/carta-backend-unit/test/TestIcd.cc:191: Failure
> Expected equality of these values:
>   _message_count
>     Which is: 1
>   3
> /home/mingyi/carta-backend-unit/test/TestIcd.cc:205: Failure
> Value of: open_file_ack.success()
>   Actual: false
> Expected: true
> /home/mingyi/carta-backend-unit/test/TestIcd.cc:220: Failure
> Expected equality of these values:
>   _message_count
>     Which is: 1
>   2
> /home/mingyi/carta-backend-unit/test/TestIcd.cc:242: Failure
> Expected equality of these values:
>   _message_count
>     Which is: 1
>   4
> /home/mingyi/carta-backend-unit/test/TestIcd.cc:264: Failure
> Expected equality of these values:
>   _message_count
>     Which is: 1
>   4
> [  FAILED  ] IcdTest.AnimatorNavigation (512 ms)
> [ RUN      ] IcdTest.AnimatorPlayback
> 2021-11-15 07:23:32	INFO	FITSCoordinateUtil::fromFITSHeader	passing empty or nonexistant spectral Coordinate axis

markccchiang commented:
> > [       OK ] IcdTest.AnimatorDataStream (456 ms)
> > [ RUN      ] IcdTest.AnimatorNavigation
> > sh: 1: fits2idia: not found
> > sh: 1: fits2idia: not found

Looks like you do not install the `fits2idia` executor?

acdo2002 commented:
@markccchiang Yes, you are right, after I install fits2idia, the (2) problem disappears!!
but (1) is still there with Ubuntu 20.04....

markccchiang committed:
exist the thread pool at the end of tests

veggiesaurus reviewed (approved):
will merge in once the CI shows all tests passing

veggiesaurus committed:
Merge branch 'dev' into mark/797_ICD_tests

veggiesaurus committed:
removed old tbb includes

markccchiang committed:
fix a typo
